File: queue/queue.c

// Queue using LinkList


#include<stdio.h>
#include<stdlib.h>

struct Node {
    int data;
    struct Node *next;
}*front=NULL, *rear=NULL; 


void enqueue(int x){

    // Checking if stack is full;
    struct Node *newNode;
    newNode=(struct Node*)malloc(sizeof(struct Node));
    if(newNode == NULL){
        printf("Stack is full\n");
    }else{

        newNode->data = x;
        newNode->next = NULL;
        //! Special case (first node)
        if(front == NULL) {
            front=rear=newNode;
        }else{
            rear->next = newNode;
            rear=newNode;
        }


    }

}

int dequeue(){
    int x=-1;
    // check if stack is empty;
    if(front == NULL){
        printf("Stack is empty\n");
    }else{
        struct Node *temp;
        temp=front;
        front=front->next;
        x=temp->data;
        free(temp);
    }
    return x;
}

void display(){

    struct Node* iterator;
    iterator=front;
    while(iterator){
        printf("%d->", iterator->data);
        iterator=iterator->next;
    }

}

int main(){
    
    int d;
    printf("ENQUEUE OPERATION\n");
    enqueue(10);
    enqueue(20);
    enqueue(30);
    display();
    printf("DEQUEUE OPERATION\n");
    d=dequeue();
    printf("ELEMENT DEQUED is %d\n",d);
    d=dequeue();
    printf("ELEMENT DEQUED is %d\n",d);
    enqueue(40);
    enqueue(50);
    display();   

}

File: queue/queue.js

class Node {
    constructor(data) {
        this.data = data;
        this.next = null;
    }
}

class Queue {
    constructor() {
        this.front = null;
        this.rear = null;
    }

    enqueue(x) {
        let t = new Node(x);
        if (!t) {
            console.log("Queue Overflow");
        } else {
            if (this.front === null) {
                this.front = t;
                this.rear = t;
            } else {
                this.rear.next = t;
                this.rear = t;
            }
        }
    }

    dequeue() {
        let x = -1;
        if (this.isEmpty()) {
            console.log("Queue Underflow");
        } else {
            let p = this.front;
            this.front = this.front.next;
            x = p.data;
        }
        return x;
    }

    isEmpty() {
        return this.front === null;
    }

    display() {
        let p = this.front;
        let result = "";
        while (p) {
            result += p.data;
            p = p.next;
            if (p) {
                result += " <- ";
            }
        }
        console.log(result);
    }
}

// Main function equivalent
let A = [1, 3, 5, 7, 9];

let q = new Queue();

for (let i = 0; i < A.length; i++) {
    q.enqueue(A[i]);
}

q.display();

for (let i = 0; i < A.length; i++) {
    q.dequeue();
}

q.dequeue();
